CTP-4382 : using is_teacher() from feedback tracker

opitz opened the pull request:
also: checking for teacher role in each enrolled course before showing
This now also contains the code for CTP-4443 (toggle between marker and student content).
It depends on https://github.com/ucl-isd/moodle-report_feedback_tracker/pull/79

leonstr commented on block_my_feedback.php:
```diff
@@ -111,8 +92,9 @@ public static function fetch_marking(stdClass $user): ?array {
         $marking = [];
```
```
 88         global $DB;
```

Global not used in function.

opitz replied:
removed

leonstr commented on block_my_feedback.php:
```diff
@@ -111,8 +92,9 @@ public static function fetch_marking(stdClass $user): ?array {
+                    !report_feedback_tracker\local\site::is_teacher($course->id)) {
                 continue;
             }
 
```
```
162         // Sort and return data.
163         if ($marking) {
164             usort($marking, function ($a, $b) {
165                 return $a->unixtimestamp <=> $b->unixtimestamp;
166             });
167 
168             return array_slice($marking, 0, 5);
169         }
170         return null;
```

Can't we just have:

```
        // Sort and return data.
        usort($marking, function ($a, $b) {
            return $a->unixtimestamp <=> $b->unixtimestamp;
        });

        return array_slice($marking, 0, 5);
```

It's three lines shorter and Mustache treats an empty array as a false value so it should have identical behaviour.

opitz replied:
updated.

leonstr commented on block_my_feedback.php:
```diff
@@ -111,8 +92,9 @@ public static function fetch_marking(stdClass $user): ?array {
-            if (!$course->visible || !self::is_course_current($course)) {
+            // Skip hidden or non-current courses or courses the user has no teacher role in.
+            if (!$course->visible || !self::is_course_current($course) ||
+                    !report_feedback_tracker\local\site::is_teacher($course->id)) {
```
This is a bit long-winded. Instead we could use a namespace alias, e.g.:

```
use report_feedback_tracker\local\site as feedback_tracker_site;
⋮
            if (!$course->visible || !self::is_course_current($course) ||
                    !feedback_tracker_site::is_teacher($course->id)) {
```

opitz replied:
refactored .

leonstr commented on block_my_feedback.php:
```diff
@@ -111,8 +92,9 @@ public static function fetch_marking(stdClass $user): ?array {
-            if (!$course->visible || !self::is_course_current($course)) {
+            // Skip hidden or non-current courses or courses the user has no teacher role in.
+            if (!$course->visible || !self::is_course_current($course) ||
+                    !report_feedback_tracker\local\site::is_teacher($course->id)) {
```
Here and in `report_feedback_tracker\local\site::get_feedback_tracker_site_data()` we pass `$course->id` to `is_teacher()` . But `is_teacher()` only uses this value to look up the course's context ID which is already known!

So instead we could have:

```
    public static function is_teacher($course = null): bool {
        ⋮
        if ($course) {
            // Check if user has expected role in the given course.
            foreach ($roles as $role) {
                if (user_has_role_assignment($USER->id, (int) $role, $course->ctxid)) {
```

and call it with:

```
                if (in_array($data->year, $courseacademicyears) &&
                        in_array($data->termcode, $courseterms[$data->year]) &&
                        self::is_teacher($course)) {
```

in `site.php` and:

```
            if (!$course->visible || !self::is_course_current($course) ||
                    !report_feedback_tracker\local\site::is_teacher($course)) {
```

in `block_my_feedback.php`.

opitz replied:
refactored in https://github.com/ucl-isd/moodle-report_feedback_tracker/pull/79

leonstr commented on block_my_feedback.php:
```diff
@@ -111,8 +92,9 @@ public static function fetch_marking(stdClass $user): ?array {
-            if (!$course->visible || !self::is_course_current($course)) {
+            // Skip hidden or non-current courses or courses the user has no teacher role in.
+            if (!$course->visible || !self::is_course_current($course) ||
+                    !report_feedback_tracker\local\site::is_teacher($course->id)) {
```
We're checking:

```
                if ($summative->cmid == 0 || !in_array($summative->cmid, $cmids)) {
```

and:

```
                if (!$mod->visible) {
```

But we could do these checks in a local_assess_type function which would make this code cleaner. So in `local/assess_type/classes/assess_type.php` have a function like:

```
    public static function get_cm_records_by_courseid(int $courseid, ?int $type = null): array {
        global $DB;
        $sql = "SELECT at.*
                  FROM {local_assess_type} at
                  JOIN {course_modules} cm ON cm.id = at.cmid
                 WHERE courseid = :courseid AND visible = 1";
        $params = ['courseid' => $courseid];

        if ($type) {
            $sql .= " AND type = :type";
            $params['type'] = $type;
        }

        return $DB->get_records_sql($sql, $params);
    }
```

Then here we just need:

```
            // Skip if no summative assessments.
            if (!$summatives = assess_type::get_cm_records_by_courseid($course->id, assess_type::ASSESS_TYPE_SUMMATIVE)) {
                continue;
            }

            // Get course mod ids.
            $modinfo = get_fast_modinfo($course->id);

            // Loop through assessments for this course.
            foreach ($summatives as $summative) {
                // Begin to build mod data for template.
                $cmid = $summative->cmid;
                $mod = $modinfo->get_cm($cmid);

                // Template.
                $assess = new stdClass;
                ⋮
```

I've gone off topic here, this probably needs its own story.

opitz replied:
Yeah, I opt in for a separate story. Feels like too many balls in the air...
Will leave it as is for now.

leonstr replied:
Agree: [CTP-4471](https://ucldata.atlassian.net/browse/CTP-4471).

leonstr commented on templates/content.mustache:
```diff
@@ -55,8 +55,8 @@
         {{#releaseddate}}
         <div class="small">
             {{#tutorname}}{{.}} · {{/tutorname}} {{releaseddate}}
```
Mix of styles here, we could have:

```
            {{#tutorname}}{{.}} · {{/tutorname}} {{.}}
```

opitz replied:
.

leonstr commented on block_my_feedback.php:
```diff
@@ -39,7 +39,7 @@ public function init() {
         if (!isset($USER->firstname)) {
             $this->title = get_string('pluginname', 'block_my_feedback');
-        } else if (self::is_teacher()) {
+        } else if (report_feedback_tracker\local\site::is_teacher()) {
```
Could use a namespace alias to make this a bit more readable:

```
use report_feedback_tracker\local\site as feedback_tracker_site;
```

opitz replied:
done.

leonstr commented on block_my_feedback.php:
```diff
@@ -63,7 +63,7 @@ public function get_content(): stdClass {
         $template = new stdClass();
 
-        if (self::is_teacher()) {
+        if (report_feedback_tracker\local\site::is_teacher()) {
```
Could use a namespace alias to make this a bit more readable:

```
use report_feedback_tracker\local\site as feedback_tracker_site;
```

opitz replied:
is_teacher() has moved from site to helper class. I have adopted the alias accordingly to feedback_tracker_helper.

Commit:
CTP-4443: toggle marker and student content
- showing link to student content from marker content if marker has a 'student' archetype as well
- showing link to marker content from student content if student has a 'teacher' or 'editingteacher' archetype as well
- using is_teacher() from report_feedback_tracker
- added new is_student() function

Commit:
CTP-4382: using is_teacher() from feedback tracker
- skip courses w/o a teacher role
- using namespace alias feedback_tracker_helper
- renamed feedback_tracker namespace alias to feedback_tracker_admin
- addressed various code review issues

opitz commented:
@leonstr I have now added the code for CTP-4443 (https://ucldata.atlassian.net/browse/CTP-4443) here as well.

stuartlamour commented on block_my_feedback.php:
```diff
@@ -104,15 +117,15 @@ public static function is_teacher(): bool {
-            // Skip hidden or non-current courses.
-            if (!$course->visible || !self::is_course_current($course)) {
+            // Skip hidden or non-current courses or courses the user has no teacher role in.
+            if (!$course->visible || !self::is_course_current($course) ||
```
In terms of coding conventions and best practice, an if statement should ideally have a single, clear condition.
If statements that contain more than 2 conditions should be broken down into more readable statements.
Refactoring this into separate statements will provide you with more readable, maintainable, and less error prone code.

opitz replied:
I think, @leonstr likes it just the opposite, but why not...
Refactored.

Commit:
CTP-4382 : splitting up if clause

Commit:
CTP-4443 : toggle marker and student view
- show a link on the marker view to student view only if user is also student and has feedback to show
- show a link on the student view to marker view only if user is also marker and has markings to show

Commit:
CTP-4382 : showing student view by default
- add icon and use Moodle strings for links,

leonstr commented on block_my_feedback.php:
```diff
@@ -123,14 +123,19 @@ public static function fetch_marking(stdClass $user): ?array {
+                    assess_type::ASSESS_TYPE_SUMMATIVE)) {
                 continue;
             }
 
```
Off topic:

```
254         if (isset($course->enddate) &&
255             $course->enddate != 0 && time() > strtotime('+3 month', $course->enddate)) {
```

I can't see that `isset($course->enddate)` can ever be false so surely we can just have:

```
        if ($course->enddate != 0 && time() > strtotime('+3 month', $course->enddate)) {
```

leonstr commented on block_my_feedback.php:
```diff
@@ -38,14 +38,14 @@ class block_my_feedback extends block_base {
         global $USER;
 
-        $studentview = optional_param('student', null, PARAM_INT);
+        $markerview = optional_param('marker', null, PARAM_INT);
```
If we're going to check the logical NOT of this, why not use false as the default:

```
        $markerview = optional_param('marker', false, PARAM_INT);
```

leonstr commented on block_my_feedback.php:
```diff
@@ -66,21 +66,22 @@ public function get_content(): stdClass {
         $template = new stdClass();
 
-        $studentview = optional_param('student', null, PARAM_INT);
+        $markerview = optional_param('marker', null, PARAM_INT);
```
If we're going to check the logical NOT of this, why not use false as the default:

```
        $markerview = optional_param('marker', false, PARAM_INT);
```

leonstr commented on block_my_feedback.php:
```diff
@@ -38,14 +38,14 @@ class block_my_feedback extends block_base {
-            $this->title = get_string('markingfor', 'block_my_feedback').' '.$USER->firstname;
-        } else {
+        } else if (self::is_student() && !$markerview) {
             $this->title = get_string('feedbackfor', 'block_my_feedback').' '.$USER->firstname;
+        } else {
+            $this->title = get_string('markingfor', 'block_my_feedback').' '.$USER->firstname;
```
Could use placeholders:

```
$string['feedbackfor'] = 'Feedback for {$a}';
⋮
$string['markingfor'] = 'Marking for {$a}';
```

```
            $this->title = get_string('feedbackfor', 'block_my_feedback', $USER->firstname);
        } else {
            $this->title = get_string('markingfor', 'block_my_feedback', $USER->firstname);
```

leonstr replied:
We're assuming if they're not a student they're a teacher. But this means that an authorised user not enrolled on any courses sees "Marking for ..." when they add the block when edit mode is enabled, which is presumably not correct.

They see nothing when edit mode is off.

leonstr commented on block_my_feedback.php:
```diff
@@ -66,21 +66,22 @@ public function get_content(): stdClass {
             // If user has a teacher role and some markings to do, show a link to the teacher content.
             if (feedback_tracker_helper::is_teacher() && self::fetch_marking($USER)) {
-                $template->markerurl = $PAGE->url;
+                $template->markerurl = $PAGE->url . '?marker=1';
```
If the user is a teacher but is enrolled in one or more courses as a student, but does not have feedback in those courses, then `$template->mods` will be null. So despite then populating `$template->markerurl` they won't see this, or see anything at all, because:

```
        if (isset($template->mods)) {
            $this->content->text = $OUTPUT->render_from_template('block_my_feedback/content', $template);
        }
```

is not executed.

leonstr commented on block_my_feedback.php:
```diff
@@ -38,14 +38,14 @@ class block_my_feedback extends block_base {
-        } else if (feedback_tracker_helper::is_teacher() && ! $studentview) {
-            $this->title = get_string('markingfor', 'block_my_feedback').' '.$USER->firstname;
-        } else {
+        } else if (self::is_student() && !$markerview) {
```
We call `is_student()` again in `get_content()`. Instead we could remember this, something like:

```
    private static bool $isstudent;
    ⋮
    public function init() {
        self::$isstudent = self::is_student();
        ⋮
        } else if (self::$isstudent && !$markerview) {
    ⋮
    public function get_content(): stdClass {
        ⋮
        if (self::$isstudent && !$markerview) {
            ⋮
            if (self::$isstudent && $this->fetch_feedback($USER)) {
```

leonstr commented on block_my_feedback.php:
```diff
@@ -68,19 +68,18 @@ public function get_content(): stdClass {
         $studentview = optional_param('student', null, PARAM_INT);
 
-        if (feedback_tracker_helper::is_teacher() && !$studentview) {
+        if (feedback_tracker_helper::is_teacher() && !$studentview && $template->mods = self::fetch_marking($USER)) {
```
It looks like we call `fetch_marking()` to see if a teacher is also a student, and `fetch_feedback()` to see if a student is also a teacher. The resulting data is discarded even though, if the user clicks the toggle link, this is the data we want to show.

Have we considered putting the results of these calls in a hidden DOM element and having the link hide/show the respective elements?

leonstr commented on templates/content.mustache:
```diff
@@ -32,15 +32,31 @@
 
 {{! Link to student view }}
+{{#sitereporturl}}
+    <div>
```
Does this `<div>` do anything? I don't think it has any effect.

leonstr commented on templates/content.mustache:
```diff
@@ -32,15 +32,31 @@
 {{#studenturl}}
-    <div><a href="{{.}}">Student view</a></div>
-    <hr>
+    <div>
```
Does this `<div>` do anything? I don't think it has any effect.

opitz commented:
Will not be implemented as the block needs it's own check for markings.
